cita-network/src/node_manager.rs

```diff
@@ -126,42 +134,116 @@ impl TransformAddr {
     }
 }
 
+#[derive(Default, Debug)]
+pub struct ConsensusNodeTopology {
+    pub linked_nodes: BTreeSet<Address>,
+    pub validator_nodes: BTreeSet<Address>,
+    //pub consensus_threshold_linked : bool,
+    pub consensus_all_linked: bool,
+    pub height: u64,
+}
+
+impl ConsensusNodeTopology {
+    pub fn new(self_address: Address) -> ConsensusNodeTopology {
+        let mut top = ConsensusNodeTopology::default();
+        top.linked_nodes.insert(self_address);
+        top
+    }
+
+    fn validator_subset_linked(&self) -> bool {
+        self.validator_nodes.is_subset(&self.linked_nodes)
+    }
+
+    pub fn update_validators(&mut self, height: u64, validators: BTreeSet<Address>) {
+        if height < self.height || validators == self.validator_nodes {
+            debug!("No need update validator height {} self height {} validator {:?} self validator {:?}",
+            height,self.height,validators,self.validator_nodes);
+
+            if height > self.height {
+                self.height = height;
+            }
+            return;
+        }
+        self.validator_nodes = validators;
+        self.consensus_all_linked = self.validator_subset_linked();
+    }
+
+    pub fn add_linked_nodes(&mut self, linked_node: Address) {
+        if self.linked_nodes.insert(linked_node) {
+            self.consensus_all_linked = self.validator_subset_linked();
+        }
+    }
+
+    pub fn del_linked_nodes(&mut self, linked_node: &Address) {
+        if self.linked_nodes.remove(linked_node) {
+            self.consensus_all_linked = self.validator_subset_linked();
+        }
+    }
+
+    pub fn consensus_all_linked(&self) -> bool {
+        self.consensus_all_linked
+    }
+    //pub fn consensus_threshold_linked(&self) -> bool {self.consensus_threshold_linked}
```

Comment:
Same.
It's better that the comment is just used for explaining the code.
The git will keep the history.